pkg/util/fileutil/writer.go

```diff
@@ -51,72 +35,58 @@ func Validate(payloads []*v1alpha1.File) error {
 	return nil
 }
 
-// WritePayloads writes the files to target directory. One of the main
-// deviations from atomic_writer.go is that this does not attempt to be atomic
-// in updates. atomic_writer.go relies upon ../data folder, but in CSI driver
-// context that path is outside of the tmpfs.
+// WritePayloads writes the files to target directory. This helper builds the
+// atomic writer and converts the v1alpha1.File proto to the FileProjection type
+// used by the atomic writer.
 func WritePayloads(path string, payloads []*v1alpha1.File) error {
-	for _, payload := range payloads {
-		content := payload.Contents
-		mode := os.FileMode(payload.Mode)
-		fullPath := filepath.Join(path, payload.Path)
-		baseDir, _ := filepath.Split(fullPath)
-
-		if err := os.MkdirAll(baseDir, os.ModePerm); err != nil {
-			return fmt.Errorf("unable to create directory %s: %w", baseDir, err)
-		}
+	// cleanup any payload paths that may have been written by a previous
+	// version of the driver/provider.
+	if err := cleanupProviderFiles(path, payloads); err != nil {
+		return err
```

Comment:
wrap this error to give more context that the issue occurred to the cleanup of existing files